---
title: "Genetic_filtering_T_Adapcon_Gentree_version_Readpos_rank_MQ"
author: "Thomas Francisco"
date: "2024-03-11"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = FALSE)

library(vcfR)
library(reshape2)
library(ggplot2)
library(SNPfiltR)
library(RColorBrewer)
library(stringr)
library(dplyr)
library(writexl)
library(tidyr)
```


This script is filtering the genetic data of Taxus baccata Adapcon-Gentree dataset and output 2 files: 
  - 1 for population structure (without mac)
  - 1 for GEA and other analysis


The raw vcf file contains 11374 SNPs, 490 individuals and 8.76 % of missing data
```{r data}
vcf <- read.vcfR("C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/genetic/filtered_T_11374SNP_490IND_ADAPCON_GENTREE.vcf", verbose = FALSE,convertNA=T)
meta_data_vcf=read.csv("C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/Samples/samples_taxus_baccata_adapcon_gentree.csv",h=T,sep=";",dec=",")
```



    1. Population structure dataset
    
Filtering of DP, allele_balance, Na_snp, Na_indiv
```{r filters, warning=FALSE}
min_dp= 7
max_dp=250
Allele_balance_min= .2
Allele_balance_max= .8
Na_snp=0.85 #% of Na per SNP,  0.85 = we kept only snp with less than 15% of missing data
Na_indiv=0.15#% of Na per individual, 0.15 = we kept only individuals with less than 15% of missing data


vcf_filter_structure <- vcf %>% 
  hard_filter(depth=min_dp ) %>% 
  max_depth(maxdepth =max_dp )%>% 
  filter_allele_balance(min.ratio=Allele_balance_min,max.ratio=Allele_balance_max)%>% 
  missing_by_snp(cutoff=Na_snp)%>%
  missing_by_sample( cutoff =Na_indiv )
```
After filtering, we have 8252 SNPs, 452 samples and 4.37% of Na

Output the results for :

   - Structure
   
We need to transform the vcfR to a file for the structure software (first row with the name of the SNP, all the other rows 1 tree per row and in columns the SNP with genotype data in two different columns for each snp: 1/1 will be 1 in one column and 1 in another)
Few steps in R then 2 out of R: 
  - vcfR data to a dataset
```{r VCFR to genind}
#change VCFR to dataframe
vcf_data <- vcf_filter_structure %>% 
  extract.gt(element="GT", IDtoRowNames = T, as.numeric=F) %>% t %>% data.frame
```

  - split the genotypic information into 2 columns for each SNP: we use the split_columns function 
  
```{r split columns}
#the function
split_columns <- function(data) {
  new_data <- data.frame(matrix(ncol = 0, nrow = nrow(data)))  # Create an empty dataframe for the output
  
  for (col in names(data)) {#apply the function for each columns of the dataframe
    split_parts <- strsplit(as.character(data[[col]]), "/") #split the columns based on /
    max_parts <- max(sapply(split_parts, length)) # Maximum number of split parts in the column -> store the new columns to then gave a name
    
    for (i in 1:max_parts) {#names the columns by the name of the snp + _part 1 or 2
      new_column_name <- paste0(col, "_part", i)
      new_data[[new_column_name]] <- sapply(split_parts, function(x) ifelse(length(x) >= i, x[i], NA))
    }
  }
  row.names(new_data)<-row.names(data)
  return(new_data)
}

#the input data
data_split <- split_columns(vcf_data)

```
Final dataset and exporting it to a txt file

```{r outputing structure dataset to excel file}

write.table(data_split, "C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/genetic_new/data_split.txt")
```
  - then we need to replace the name of the snp to have only 1 nma efor each snp and not 2 for the structure input: we download the snp name from the vcf data:
```{r name of the SNPs}
write_xlsx(vcf_data,"C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/genetic_new/nom_snp.xlsx")
```
  and also need to replace Na to -9 and delete the "" from the txt

  - finally using ubuntu, we need to modify the txt to fit the txt format of linux because we created the file in windows and will be used by a linux software. We used:  dos2unix nameofdata.txt
  

  
   - PCA

We are now changing the format of the vcf to fit the format for pca analysis
First, we're keeping only the genotype information from the vcfR

```{r extracting genotypic informations}
vcf_data <- vcf_filter_structure %>% 
  extract.gt(element="GT", IDtoRowNames = T, as.numeric=F) %>% t %>% data.frame
```

We also need to change the format of the dataset from nucleotide information to genotype 
```{r changing format nucleotide to genotype, warning=FALSE}
#the format is in 0/0, 0/1 and 1/1 so we transform this in allele count

genetic_data_geno <- vcf_data %>% 
  mutate_all(funs(str_replace(., "0/0", "0"))) %>% 
  mutate_all(funs(str_replace(., "0/1", "1"))) %>% 
  mutate_all(funs(str_replace(., "1/1", "2")))
```

We want to add to the file, the % of Na per individuals to plot them on the PCA.
```{r calculation of the Na per individuals}
#we need to specify that SNPs are numeric to calculate the Na per individuals 
Dataset_PCA_8252SNP_452IND <- data.frame(row.names(vcf_data),apply(genetic_data_geno, 2, as.numeric))

#calculation of % NA
na_percentage_indiv <- rowMeans(is.na(Dataset_PCA_8252SNP_452IND[,-c(1)])) * 100
```


```{r output dataset}

Dataset_PCA_8252SNP_452IND$na_percentage_indiv <- na_percentage_indiv
names(Dataset_PCA_8252SNP_452IND)[names(Dataset_PCA_8252SNP_452IND) == 'row.names.vcf_data.'] <- 'VCF_ID'


#save the genetic dataset 
save(Dataset_PCA_8252SNP_452IND,file="C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/genetic_new/structure/Dataset_PCA_8252SNP_452IND.Rdata")

```


    2. GEA and other analysis
    
We perform again the filtering with the MAC this time:
```{r filtering for GEA}

min_dp= 7
max_dp=250
Allele_balance_min= .2
Allele_balance_max= .8
mac=20
Na_snp=0.7 #% of Na per SNP,  0.85 = we kept only snp with less than 15% of missing data
Na_indiv=0.3#% of Na per individual, 0.15 = we kept only individuals with less than 15% of missing data


vcf_filter_GEA <- vcf %>% 
  hard_filter(depth= min_dp) %>% 
   max_depth(maxdepth =max_dp )%>%
  filter_allele_balance(min.ratio=Allele_balance_min,max.ratio=Allele_balance_max) %>% 
  min_mac(min.mac = mac) %>% 
  missing_by_snp(cutoff=Na_snp) %>%
  missing_by_sample( cutoff = Na_indiv)

```
After filtering, we have 6693 SNPs, 451 samples and 4.43% of Na 

We are now changing the format of the vcf to fit the format for GEA analysis
First, we're keeping only the genotype information from the vcfR

```{r extracting genotypic informations}
vcf_data_GEA <- vcf_filter_GEA %>% 
  extract.gt(element="GT", IDtoRowNames = T, as.numeric=F) %>% t %>% data.frame
```

We also need to change the format of the dataset from nucleotide information to genotype 
```{r changing format nucleotide to genotype, warning=FALSE}
#the format is in 0/0, 0/1 and 1/1 so we transform this in allele count

genetic_data_geno <- vcf_data_GEA %>% 
  mutate_all(funs(str_replace(., "0/0", "0"))) %>% 
  mutate_all(funs(str_replace(., "0/1", "1"))) %>% 
  mutate_all(funs(str_replace(., "1/1", "2")))
```


Then, we performed the imputation because GEA do not support missing data. We will impute based on the main gene pools identified in STRUCTURE analysis. 
The imputation is based on the most common genotype per main gene pools
```{r IMPUTATION}

#Load the metadata with the main gene pools 
meta_data_vcf=read.csv("C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/Samples/samples_taxus_baccata_adapcon_gentree.csv",h=T,sep=";",dec=",")

#Keep ID info
vcf_filtered <- genetic_data_geno %>% tibble::rownames_to_column(var = "VCF_ID")#we put the ID of samples in a column
list_indiv_427 <- list(vcf_filtered$VCF_ID)

#create a metadata keeping only individuals retained after filtering steps
meta_data_vcf_imputation <- meta_data_vcf[meta_data_vcf$VCF_ID %in% vcf_filtered$VCF_ID,] 

# Group by cluster_names
gen_grouped <- vcf_filtered %>%
dplyr::group_by(meta_data_vcf_imputation$group_imput)
#we need to specify that SNPs are numeric
gen_grouped_num <- data.frame(gen_grouped[,c("VCF_ID","meta_data_vcf_imputation$group_imput")],apply(gen_grouped[,c(2:8873)], 2, as.numeric))
#we created a function to calculate the most common genotype
mostC_geno <- function(x) {
names(which.max(table(x)))
}

#we calculated the most common genotype for each SNPs
most_common <- gen_grouped_num %>%
summarise(across(-VCF_ID, ~mostC_geno(.)))

# replace missing values for each snp with the most common geno for that snp
gen_imp <- gen_grouped_num %>%
group_by(meta_data_vcf_imputation$group_imput) %>%
mutate(across(-VCF_ID, ~ifelse(is.na(.), most_common[[as.character(cur_column())]], .))) %>% 
ungroup()

# Convert ID column back to row names
rownames(gen_imp) <- gen_imp$VCF_ID
vcf_filtered_inputed <- gen_imp[,-c(1,2)] # remove ID column from data frame
rownames(vcf_filtered_inputed) <- gen_imp$VCF_ID

# Check the proportion of missing data before and after imputation
prop.table(table(is.na(vcf_filtered)))
prop.table(table(is.na(vcf_filtered_inputed)))

#add NA percentage
na_percentage_indiv <- rowMeans(is.na(gen_grouped_num[,-c(1,2)])) * 100
vcf_filtered_T_GEA_SNP_IND <- data.frame(gen_imp[,-c(1,2,7549)]);vcf_filtered_T_GEA_SNP_IND$na_percentage <- na_percentage_indiv;rownames(vcf_filtered_T_GEA_SNP_IND) <- gen_grouped_num$VCF_ID

```
We will save 2 types of genomic data for further analysis: 
      - individual-level genomic data
      - population-level genomic data
Both could be used as input of GEA or other genomic analysis. 

    1. Individual-level genomic data
This is basically the output of the imputation steps

```{r save genomic data}
#save the dataframes to load it in further scripts to avoid going through this script every time
save(vcf_filtered_T_GEA_SNP_IND,file="C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/genetic_new/vcf_filtered_T_GEA_SNP_IND.Rdata")
```

    2. Population-level genomic data
To work at the population-level, we need to calculate allelic frequencies for each snp per population. 


```{r initialisation}

#add VCF_ID but normally it will be in the imputated file
genetic_data_geno$VCF_ID <- row.names(genetic_data_geno)
#reduce the number of indiv in metadatavcf but normally it will be already done
meta_data_vcf_452 <- meta_data_vcf[meta_data_vcf$VCF_ID %in% genetic_data_geno$VCF_ID,] 


#data for allelic frequencies 
data_frequencies <-mutate(meta_data_vcf_452,genetic_data_geno,VCF_ID)

```

```{r calculation of allelic frequencies}
#formatting the genomic data
data_frequencies_num <- data_frequencies[,-c(1:3)] %>% #keeping only the snps
  apply(2,as.numeric) /2 #we divided by 2 because of the format of genomic data: 0,1,2 and for allelic frequencies we want 0,0.5, 1

#dataset with all information and genomic data in the right format
data_frequencies_num_tot <- data.frame(data_frequencies[,c(1:3)],data_frequencies_num)

#calculation of allelic frequencies
allelic_frequencies <-data_frequencies_num_tot %>% select(-c("VCF_ID","Country","Population")) %>% #remove non genomic data from the dataset
  group_by(data_frequencies_num_tot$Population) %>% #we want the allelic frequencies at the population level so we grouped
  summarise_at(vars(everything()),funs(mean),na.rm=T) %>% #calculate the mean for each snp per pop
  ungroup() %>%
as.data.frame()

```

Save the dataset of allelic frequencies

```{r save alellic frequencies dataset}
#move population level to row.names
data_allelic_frequencies_29pop_adapcon_gentree <- allelic_frequencies %>% select(-c("data_frequencies_num_tot$Population")); row.names(data_allelic_frequencies_29pop_adapcon_gentree) <- allelic_frequencies$'data_frequencies_num_tot$Population'

#save

save(data_allelic_frequencies_29pop_adapcon_gentree,file="C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/genetic_new/data_allelic_frequencies_29pop_adapcon_gentree.Rdata")

```

