---
title: "LFMM_outliers_detection"
author: "Thomas Francisco"
date: "2024-04-10"
output: html_document
---

```{r packages}
knitr::opts_chunk$set(cache = FALSE)

rm(list = ls())
#if(!requireNamespace("qvalue", quietly = TRUE)) {  
 # if (!requireNamespace("BiocManager", quietly = TRUE))
  #  install.packages("BiocManager")
  #BiocManager::install(version = "3.14")
  #BiocManager::install("qvalue")
#}
#if(!requireNamespace("lfmm", quietly = TRUE)) {  
 # remotes::install_github("bcm-uga/lfmm")
#}
library(lfmm)     # Used to run LFMM
library(qvalue)
library(LEA)
library(ggplot2)
library(writexl)
library(DescTools)#change columns, rows ...
library(magrittr)#change columns, rows ...
```


This script will perform the outliers detection using the LFMM methods (Frichot et al. 2013). This script is based on the LFMM tutorial (Frichot and François) and the tutorial of Brenna R. Forester. 

LFMM is a method that can be univariate or multivariate because we could calculate the association of SNPs with explanatory variables one by one or with all the explanatory variables together. This method can correct for population structure or not. This in a linear methods because it assumes a linear relationship between loci and climatic variables.
LFMMs are regression models combining fixed (cliamtic variables) and latent effects (parameters influencing the association like population structure) (Caye et al. (2019)). 
LfMM use a framework of mixte model with latent factor, Latent factor are hidden factors that represent background residual levels of
population structure. 
Latent factor are calulating using factorization methods -> that is really close to the method used in PCA -> explain why results of latent factor are close to PCs axis Frichot et al 2013.
How this work in term of models, the effect of the environmental variables (primary parameters) on the response variables is taking into account while correcting for the population structure -> the position of the individuals or population in a space just like a PCA space if I simplify it. The more the populations are close in the space the more their score will be similar and can explain the genetic variation (because the more the population are closed the more we expect them to have genetic compo) and more the population are far the less their score will explain the genetic variation and then other factor could explain the genetic vatiation. (it's like the example of the weigh explain by the heigh and the sex, regardless of the heigh, we found that the sex has an impact on the weigh, we study the effect of the heigh on the weigh when the sex is constant(we remove the effect of the sex on the weigh by applying is coeeficient) and same things when we study the sex, we study this when the heigh is constant)


There is two types of LFMM analys: 
- lfmm()  is Bayesian method that uses a Monte-Carlo Markov Chain algoritm
- lfmm2() is a frequentist approach that uses least-squares estimates (better for large dataset 1000 to 10000 because it's faster and also because with large dataset there is a convergence in results between frequentist and bayesian analysis).

LFMM is an analysis using allele count so we need to perform this analysis at the individual level. 

We will use the multivariate LFMM analysis. 
There is few steps: 
      - Load genetic and climatic data
      - Estimate the number of latent factor and run LFMM
      - Calculate the pvalues and Check the pvalues distribution and correct by the GIF if necessary
      - Select the threshold for outliers
      

    1. Load genetic and climatic data
    
  
To perform LFMM, genomic data needs to be in alleles count so at the individual level and not population level.  
```{r load data}

#climatic data at individual level
climatic_data <- read.csv("C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/Climatic_data/corr0.75/climatic_data_indivdual_level.csv",sep=";",dec=",")

#genomic data alleles count
load("C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/genetic_new/Gen_matrix_imp_T_Adapcon_Gentree_475_8616.Rdata")
genomic_data <- Gen_matrix_imp_T_Adapcon_Gentree_475_8616

#meta_data
meta_data_pop <- read.csv("C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/Populations/taxus_sample_29pop.csv",h=T,sep=";",dec=",")
```

We need to adjust climatic data to the number of individuals in genomic data
```{r climatic data}

climatic_data_475 <- climatic_data[climatic_data$VCF_ID %in% row.names(Gen_matrix_imp_T_Adapcon_Gentree_475_8616),]

```


We need to transform the genomic dataset into a lfmm format 

```{r genomic data into LFMM format}

write.lfmm(genomic_data,"C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/genetic_new/GEA/genomic_data.lfmm")#here the input of write.geno is a dataset with individuals in rows, snps in columns and Na in 9

#See if the lfmm is in the right format
test <- data.frame(read.lfmm("C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/genetic_new/genomic_data.lfmm"))
```


    2. Estimate the number of latent factor and run LFMM

Based on the PC analysis with genetic data, it's seems that the 2 first PCs axis could be enough to account for population structure because we could discriminate the 3 main genes pools identified by STRUCTURE along the 2 PCs.
Latent factor is not working exactly like a PCs so we will perform the latent factor analysis to see how many latent factors are required to discriminate the 3 main genes pools. 
Moreover, "K may be replaced by estimates of population genetic structure obtained with clustering algorithms like STRUCTURE" according to Pritchard et al. 2000

```{r latent factor}

Run_LFMM2 <- lfmm2(input = "C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/genetic_new/GEA/genomic_data.lfmm", env = climatic_data_475[,-c(1:3)], K = 2, effect.sizes = T)
# GEA significance test

# showing the K = 2 estimated factors
score_LFMM <- data.frame(Run_LFMM2@U)

score_Pca <- data.frame(score_LFMM, row.names (genomic_data))
      colnames(score_Pca) <- c(paste0("PC1"), paste0("PC2"), "VCF_ID")
      #add pop info on the score_pca
  score_Pca_meta_data <- merge(climatic_data_475[,c(1:2)],score_Pca, "VCF_ID")
      
      # add country, population information
      PCa_df_score <- merge(score_Pca_meta_data, meta_data_pop, "Population")
      
      # genetic PCA
      latent_factor_LFMM2_outliers_selection <- ggplot() +
        geom_point(data = PCa_df_score, aes(PC1,PC2, color = Country)) +
        scale_colour_manual(name = "Countries",
                            values = c("orangered3", "gold2", "darkorchid3", "navyblue", "turquoise2", "green3", "blue", "red", "black", "gray", "orange", "darkgreen")) +
  xlab(paste0("PC1")) + 
  ylab(paste0("PC2")) +
  facet_wrap(~"Principa component analysis using latent factor from LFMM") +
  theme_bw(base_size = 11) +
  theme(panel.grid = element_blank(), plot.background = element_blank(), panel.background = element_blank(), strip.text = element_text(size=11))

print(latent_factor_LFMM2_outliers_selection)
      
 #save
png("C:/Users/tfrancisco/Documents/Thèse/Results/species/taxus/GEA/LFMM/latent_factor_LFMM2_outliers_selection.png");print(latent_factor_LFMM2_outliers_selection);dev.off()

```
Comparison between the genetic PC and the latent factor with k=2 seems very similar. Based on that and a discussion with Benjamin Dauphin (he told that we should take k= number of groups-1), With 3 main gene pool groups, we will retain k=2. 


    3. Calculate the pvalues and Check the pvalues distribution and correct by the GIF if necessary

```{r calculate the pvalues}
#non correcting for GIF
pv_non_corrected <- lfmm2.test(object = Run_LFMM2,
input = "C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/genetic_new/genomic_data.lfmm",
env = climatic_data_475[,-c(1:3)],
full = T,
genomic.control = F)


#correcting for GIF
pv_corrected <- lfmm2.test(object = Run_LFMM2,
input = "C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/genetic_new/genomic_data.lfmm",
env = climatic_data_475[,-c(1:3)],
full = T,
genomic.control = T)
```

The next step is to visualize the pvalue distribution with pvalues corrected or not by the genomic inflation factor."The genomic inflation factor expresses the deviation of the distribution of the observed test statistic compared to the distribution of the expected test statistic" Van den Berg et al. 2019
-> high GIF = indicate that we did not correct enough for popualtion stratification etc -> high rate of false positive
-> low GIF = we corrected too much and we lose a part of the signal -> high rate of false negatives
-> GIF near 1 = the correction is good

To visualize which GIF seems pertinent according to our data, we can look at the shape of the distribution to visualize if we have corrected enough for the population structure because we should expect a pic of pvalues near 0 and then a continuous number of pvalues from 0.1 to 1. 
We can compared the distribution of corrected and not corrected for GIF. 
```{r visualize the pvalue distribution}
#non corrected for GIF
Histogram_of_non_calibrated_Pvalues_LFMM<- hist(pv_non_corrected$pvalues,
     main= "Histogram of non-calibrated P-values",
     xlab= "P-values")

#save
png("C:/Users/tfrancisco/Documents/Thèse/Results/species/taxus/GEA/LFMM/Histogram_of_non_calibrated_Pvalues_LFMM.png");hist(pv_non_corrected$pvalues,
     main= "Histogram of non-calibrated P-values",
     xlab= "P-values"); dev.off()

#correcting for GIF
Histogram_of_calibrated_Pvalues_LFMM<-hist(pv_corrected$pvalues,
     main= "Histogram of calibrated P-values",
     xlab= "P-values")

#save
png("C:/Users/tfrancisco/Documents/Thèse/Results/species/taxus/GEA/LFMM/Histogram_of_calibrated_Pvalues_LFMM.png");hist(pv_corrected$pvalues,
     main= "Histogram of calibrated P-values",
     xlab= "P-values");dev.off()

```


Can we select our-self the GIF values? for LFMM2 univariate method (full= F), yes we can select a GIF values for each climatic variables but for the full model i don't find how to do it. 
Arguments to use the univariate vs multivariate models: 

  - univariate: less complicated models, less subject to violate the assumptions, for easy to interpret ate 
  - multivariate: take into account the complex effect or interaction of climatic variables on the genetic variations, reduce the number of test performed?

    4. Selecting the threshold for outliers






```{r threshold outlier detection}

df_pvalues_calibrated <- data.frame(SNP=colnames(genomic_data),pvalues=pv_corrected$pvalues)

#FDR correction
outliers_FDR5perc <- data.frame(snp_names=colnames(genomic_data) ,qvalues=qvalue(pv_corrected$pvalues)$qvalues)

#threshold
thres_FDR <- 0.05

outliers_T_adapcon_gentree_LFMM <- data.frame(SNP=outliers_FDR5perc$snp_names[which(outliers_FDR5perc$qvalues<thres_FDR)],qvalues = outliers_FDR5perc$qvalues[which(outliers_FDR5perc$qvalues<thres_FDR)])

length(which(outliers_FDR5perc$qvalues < thres_FDR)) ## h.w many SNPs have an FDR < 5%?


```

```{r Manhattan plot }
#selection of the outliers from FDR 5%
df_pvalues_calibrated$type <- "Neutral" ;df_pvalues_calibrated$type[df_pvalues_calibrated$SNP%in%outliers_T_adapcon_gentree_LFMM$SNP] <- "Outlier"
df_pvalues_calibrated$type <- as.factor(df_pvalues_calibrated$type)

#Bonferroni threshold
threshold_bonferroni <- 0.05/nrow(df_pvalues_calibrated)

#plot
 Manhattan_plot_LFMM_FDR_5perc<-ggplot(data = df_pvalues_calibrated_f) +
  geom_point(aes(x=1:nrow(df_pvalues_calibrated), y=-log10(pvalues), col = Outliers), size=1.4) +
  scale_color_manual(values = c("lightblue", "orange")) +
  xlab("Loci") + ylab("-log10(p.values)") +
  geom_hline(yintercept=-log10(threshold_bonferroni), linetype="dashed", color = "red", size=0.6) +
  ggtitle("Manhattan plot LFMM, with FDR 5% threshold") +
  guides(color=guide_legend(title="Type")) +
  theme_bw(base_size = 11) +
  theme(legend.position="right", panel.grid = element_blank(), strip.text = element_text(size=11),plot.title = element_text(hjust = 0.5,color = "Black",face="italic"))


print(Manhattan_plot_LFMM_FDR_5perc)
      
 #save
png("C:/Users/tfrancisco/Documents/Thèse/Results/species/taxus/GEA/LFMM/Manhattan_plot_LFMM_FDR_5perc.png");print(Manhattan_plot_LFMM_FDR_5perc);dev.off()

```


We need to save the outliers identified for downstream analysis

```{r save outliers set}
write_xlsx(outliers_T_adapcon_gentree_LFMM,"C:/Users/tfrancisco/Documents/Thèse/Results/species/taxus/GEA/outliers/outliers_T_adapcon_gentree_LFMM.xlsx")
save(outliers_T_adapcon_gentree_LFMM, file="C:/Users/tfrancisco/Documents/Thèse/Results/species/taxus/GEA/outliers/outliers_T_adapcon_gentree_LFMM.Rdata")
```



############################## Corrected matrix for Population structure for GF analysis  #######################

In order to perform a outliers detection corrected for population structure using GF, we need to use as input a genomic matrix corrected for population structure because GF cannot correct for population structure. To calculate this corrected matrix we followed Archambeau et al. 2024 using the LFMM approach to calculate the corrected geno matrix. 
To understand, here the model applied by the lfmm2 function:

where **B**, **U** and **V** are the effect size, and factor and loading matrices adjusted by the lfmm2 algorithm from the set of current environmental variables included in the matrix **X**. **B** is a matrix of dimension **p × b** with **p** the number of genetic markers and **b** the number of environmental variables. **U** is a matrix of dimension **n x K** , with **n** the number of individuals (ie genotypes) and  **K** the number of latent factors. **V** is a matrix of dimension **p x K**. **X** is a matrix of dimension ** n x b**. **Yfut** is a matrix of dimension **n x p**

We want a matrix of allele frequencies corrected for population structure, so basically:
                        Ycorrected = Yfut - UVt = XBt


Below, we do the matrix multiplication of the matrix X (dimension n x b) and the transpose of the matrix B (dimension b x p) to obtain the matrix Ycorrect (dimension n x p).

The matrix B is the matrix from the lfmm2 output
```{r correctd matrix}
# matrix X where we x by the matrix B. 
Genomic_matrix_corrected_from_LFMM_T_adapcon_gentree <- as.matrix(climatic_data_475[,-c(1:3)]) %*% t(Run_LFMM2@B) %>% set_rownames(row.names(genomic_data)) %>% set_colnames(colnames(genomic_data)) %>% as.data.frame()

```

-> The correct matrix gave very similar values of individuals of the same population because they are closed in the PCA. We have more important differences for individuals far from each other in the PCA (realized by LFMM). 

We save the corrected genotypic matrix for BF analysis
```{r save the corrected genotypic matrix}

save(Genomic_matrix_corrected_from_LFMM_T_adapcon_gentree, file="C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/genetic_new/GEA/Genomic_matrix_corrected_from_LFMM_T_adapcon_gentree.Rdata")

```


