---
title: "Processing_climatic_data_T_adapcon_Gentree_bis"
author: "Thomas Francisco"
date: "2024-02-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = FALSE)
library(dplyr) 
library(tidyverse)
library(ggplot2)
library(corrplot)  
```


This script is a pipeline to process the climatic/environmental data from the row output of *ClimateDT* (or any program that would give data in an excel with one row per populatiob/individual per year of climatic data) to a dataframe with center and reduce the choosen variables for past/present and future climatic data.
This script is separated in multiples steps:
- Pre-selection of climatic variables 
- Loading of past and future data and processing :mean values per period, check values between period and check correlation between values from point to values from raster
- Selection of climatic variables: calculation of distance between past and future values, imprecision along studied area, multicollinearity
- Comparison of the past retained climatic values between populations (see if adaptive gradient could emerge from )
- Comparison of the values of the different climatic models for the future between populations 
- Standardization of the retained climatic values


    **1. Pre-selection of climatic variables**
The first step consist of go through the literature of the targeted species and define a set of climatic variables potentially drivers (inducing selective pressure) of adaptive genetic variation across the species range.  
The goal is to retain biological informative climatic variables only. This enable to reduce the set of tested climatic variables and also enable to work on the hypothesis before starting the analysis.
The goal is not to retain only 2 or 3 variables but also not to retain 40 of them. This pre-selection could retain 10-15 variables depending on the species and the richness of the literature related to it.

**For Taxus, we retained as a pre selection: Bio1,2,4,7,12,15,AHM, precip June_July, precip Feb_to_July, Temperature June to September, Min temperature Novembre to January**

    **2. Loading climatic data and processing them**
Climatic data come from **ClimateDT** in two different types: layer of points and raster.
We extracted the data from the past and from the future. One major differences between past and future data results in the shape of the data: 
  - Past data are available in 1 value per year for each variables (for each populations/pixel selected). As a example, for the period 1901-1950, we have 50 values for each variables for each populations
  - Future data are not so accurate, we only have the mean of the values for the selected period. As a example, for the period 2041-2070, we have 1 value for each variables for each populations. -> Not sure about this because I have 1 value per year per pop even for the future data...
  
One of the issues is that all the bioclimatic variables (bio) or other related climatic variables are build using monthly values of tmin, tmax and precipitation. We cannot just perform the mean of the 50 values of bio 1 from 1901 to 1950 to obtain the mean value of bio 1 for this period. We need to average the tmin, tmax and precipitation for each month to have 1 value of tmin for January for the period 1901-1950, 1 value of tmin for February for the period 1901-1950 ... 
Then, we can use the **biovars** function from the **Dismo** R package to calculate the 19 bioclim variables for the period of interest (e.g 1901-1950).
To calculate the other climatic variables such as AHM for the same specific period, we can follow the procedure of ClimateDT by using the new Bio1,Bio10 and Bio12... obtained from biovars.
For other variables not directly linked to tmin, tmax, precipitation or bioclim variables such as GDD5, the average value for a specific period is not straightforward to get, data are coming from Maurizio Marchi (IBBR-CNR).


      - Average values for the specific periods
First of all, we load the tmin, tmax and preci data for the 2 periods: 1901-1950 and 1951-1990.
Are we loading layer point dataset and raster?

```{r loading past climatic data layer points}
Pre_past_climatic_data_point <- read.csv("C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/Climatic_data/Pre_climatic_data_29_pop_T.csv",h=T,sep=";",dec=",")

```

We need to split the dataframe to have 1 per period
```{r split the pre dataframe based on the period}
Pre_past_data_1901_1950 <- Pre_past_climatic_data_point[Pre_past_climatic_data_point$Year < 1951,]
Pre_past_data_1951_1990 <- Pre_past_climatic_data_point[Pre_past_climatic_data_point$Year >= 1951,]

```


This function uses as input a data frame with at least the population or ID, prec, tmn and tmx data for each months as columns and populations/individuals as rows. The data are: 
    - ether sum over the period(mean=T) or not (mean=F)
 **if the average is not done, you will also need to provide the position of the prec, tmn and tmx in the  input file (from_variable, to variable) and the ID/populations names (ID)**
    - bio and other variables (like AHM) are already in the input file (bio=T), or not (bio=F)
      If bio=T, then the selected_bio parameters should include all the variables already in the input file that wants to be in the output file (add like AHM...)
The output is a data frame with all the bio climatic variables (selected_bio) and other climatic variables (selected_calculated) selected. 
```{r function to calculate and compute all the climatic variables that we want in one table}
computation_climatic_variables <- function(data,selected_bio,selected_calculated,mean,bio,from_variable,to_variable,ID){

if(mean==F){##Meaning that the function need to average the tmx, tmn and prec values over the period -> layer data
  
  #create a table with only the climatic variables of interest
  climatic_data_filtered <- data.frame(data) %>% 
    mutate(across(from_variable:to_variable, as.numeric))#pass the variables in numeric
#do the mean of each year of data for each climatic variables and that's for each ID (pop)
  climatic_data_filtered_mean <- climatic_data_filtered %>% 
   group_by(ID) %>% 
  summarize(across((from_variable-1):(to_variable-1), ~ mean(.))) ##-1 because, ID is no longer a columns but a column names so it does not count it as a column
  #we need to extract the ID, longi and lati from the dataframe with all the rows
  climatic_data_mean <- climatic_data_filtered %>% 
    group_by(ID) %>% #groups the row by ID
    slice_head(n = 1) %>% #keep only the first row of each groups
    dplyr::select(c(1:3))#select only the 3 first columns
  #add the longitude and latitude variables
   Mean_pre_variables <- data.frame(climatic_data_mean[,c(1:3)],climatic_data_filtered_mean[,-1]) #-1 because ID is already in

   data_climatic <- data.frame(Mean_pre_variables) }
  
  else{#the average is already did -> raster data
    data_climatic <- data.frame(data)
  }
   
  #we extracted the prec, tmn and tmx from the inpute data
prec <- data.frame(t(data_climatic %>% dplyr::select(starts_with("prc"))))
tmn <- data.frame(t(data_climatic %>% dplyr::select(starts_with("tmn"))))
tmx <- data.frame(t(data_climatic %>% dplyr::select(starts_with("tmx"))))
  

#we calculated the average temperature for each month over the period, we will use it to calculated the mean summer_temp
tavg <- (data_climatic %>% dplyr::select(contains("tmn")) + data_climatic %>% dplyr::select(contains("tmx")))  / 2
   

if(bio==F){#Meaning that the bioclimatic variables are not in the input data and the function needs to calculate them -> layer dataset
  ##bio calculation
Bio_data <- data.frame(biovars(prec,tmn,tmx))

##we also calculate the AHM here because if bio are calculated then AHM will also be

#AHM calculation
AHM <- data.frame((Bio_data$bio1+10)/(Bio_data$bio12/1000)) #the more we are close to 0 the less it's arid, the more we are far from 0 the more it's arid
  
}

##########other indexes calculated regardless of the dataset (raster or layer)

#mean Temperature june to september
Mean_summer_temp <- data.frame(tavg$tmn06+tavg$tmn07+tavg$tmn08+tavg$tmn09)/4
#min temperature winter
tmn_cal <- data.frame(t(tmn))
Min_winter_temp <- data.frame(tmn_cal$tmn11+tmn_cal$tmn12+tmn_cal$tmn01+tmn_cal$tmn02)/4
#summer precipitation
prec_cal <- data.frame(t(prec))
Mean_summer_prec <- data.frame(prec_cal$prc06+prec_cal$prc07+prec_cal$prc08+prec_cal$prc09)/4
#Feb_spring/summer precipitation
Mean_Feb_spring_summer_prec <- data.frame(prec_cal$prc02+prec_cal$prc03+prec_cal$prc04+prec_cal$prc05+prec_cal$prc06+prec_cal$prc07+prec_cal$prc08+prec_cal$prc09)/8

##final dataset combining all the wanted climatic variables

#AHM is already in the input file of raster so we separate the final dataset whether the input file is coming from layer or raster 

if(bio==F){ #layer
  
  #we created a data frame with all the calculated variables (except the bio), then we can also choose which one we want to have in the final dataset
data_calculated_var <- data.frame(Mean_summer_temp,Min_winter_temp,Mean_Feb_spring_summer_prec,Mean_summer_prec,AHM);colnames(data_calculated_var)=c("MeanT_summer","MinT_winter","MeanP_spring_summer","MeanP_summer","AHM")

#final dataset
  climatic_data <- data.frame(cbind(Mean_pre_variables[,c("ID")],Bio_data[,selected_bio],data_calculated_var[,selected_calculated]));colnames(climatic_data)=c("Population",selected_bio,selected_calculated)
  }

else{ #raster
  
  #we created a data frame with all the calculated variables (except the bio), then we can also choose which one we want to have in the final dataset
data_calculated_var <- data.frame(Mean_summer_temp,Min_winter_temp,Mean_Feb_spring_summer_prec,Mean_summer_prec);colnames(data_calculated_var)=c("MeanT_summer","MinT_winter","MeanP_spring_summer","MeanP_summer")

  climatic_data <- data.frame(cbind(data[,c("Population")],data[,selected_bio],data_calculated_var[,selected_calculated]));colnames(climatic_data)=c("Population",selected_bio,selected_calculated)
  
}

return(climatic_data) }
```

```{r computation_climatic_variables for the 1901-1950 period}
data_layer <- Pre_past_data_1901_1950 #input file
ID_layer <- Pre_past_data_1901_1950$ID#id to calculate the average
from_variable <- 6 #position in the input file of the prec,tmn and tmx variables
to_variable <- 41  #position in the input file of the prec,tmn and tmx variables
selected_bio_layer <- c("bio1","bio2","bio4","bio7","bio12","bio15")#all the variables that have been calculated by the function as bio and need to be in the output file
selected_calculated_layer <- c("MeanT_summer","MinT_winter","MeanP_spring_summer","MeanP_summer","AHM")#all the variables that are calculated by the function and need to be in the output file

past_climatic_data_1901_1950 <- computation_climatic_variables(data_layer,selected_bio_layer,selected_calculated_layer,mean=F,bio=F,6,41,ID_layer)
```


```{r computation_climatic_variables for the 1951-1990 period}
data_layer <- Pre_past_data_1951_1990 #input file
ID_layer <- Pre_past_data_1951_1990$ID#id to calculate the average
from_variable <- 6 #position in the input file of the prec,tmn and tmx variables
to_variable <- 41  #position in the input file of the prec,tmn and tmx variables
selected_bio_layer <- c("bio1","bio2","bio4","bio7","bio12","bio15")#all the variables that have been calculated by the function as bio and need to be in the output file
selected_calculated_layer <- c("MeanT_summer","MinT_winter","MeanP_spring_summer","MeanP_summer","AHM")#all the variables that are calculated by the function and need to be in the output file

past_climatic_data_1951_1990 <- computation_climatic_variables(data_layer,selected_bio_layer,selected_calculated_layer,mean=F,bio=F,6,41,ID_layer)
```


      - Selection of the past period
We want that the past climatic climat caracterise the time of establishement of the trees and then caracterize the selective(s) pressure(s) that selected the trees with the best phenotype and so genotype for these pressure. By doing that, the GEA relationship will assume to be able to capture the correlation between the climate and the genetic variation to try to find adaptive loci or near adaptive regions. 
As lifespawn of trees is important, we assumes that the time of establishement of the trees in the populations happenned before the global warming. We selected 2 periods that could be of interest: 
    - 1901-1950: We can argue that the global warming had not impact the climate of this period, but because we went back more in the past, the accuracy of the data is reduce.
    - 1951-1990: the accuracy of these data is greater because we have more climatic station that measure this climate but we can also expect that global warming could already be visible during this period.

We want to study the correlation between these two period to see if we could detect trends of climate change in the 1951-1990 period compared to 1901-1950 through for example warming or reduction of precipitation in some areas. If not, we could then use this period as past data for GEA.

```{r comparison between the two past periods}
#we want to compare them using the difference between both period using a 1:1 relation
#first we want to color the points based on their origin: country or area, we will extract this information from another data.frame
meta_data <- read.csv("C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/Populations/taxus_sample_29pop.csv",h=T,sep=";",dec=",")
##period 1901_1950
comp_past_climatic_data_1901_1950 <- merge(meta_data[,c(1:2)],past_climatic_data_1901_1950,"Population")
colnames(comp_past_climatic_data_1901_1950)[3:13] <- paste0(colnames(comp_past_climatic_data_1901_1950)[3:13], "_1901_1950")
##period 1951_1990
comp_past_climatic_data_1951_1990 <- merge(meta_data[,c(1:2)],past_climatic_data_1951_1990,"Population")
colnames(comp_past_climatic_data_1951_1990)[3:13] <- paste0(colnames(comp_past_climatic_data_1951_1990)[3:13], "_1951_1990")


#general data frame with the 2 periods
data_tot_period <- data.frame(merge(comp_past_climatic_data_1901_1950,comp_past_climatic_data_1951_1990,"Population"))%>% 
  dplyr::select(-c("Country.y")) 

```

In order to compare the climatic variables between the two periods, we will perform a scatterplot between the both period for each climatic variables. We will be able to see if the values are consistent across periods and also the differences between and within populations along the climatic variables.

```{r function to compare the two periods}
#we created a function to perform the scatterplot 
##function
scatterplot_funct <- function(data,x,y,Countries,xlab,ylab,title){
ggplot(data,aes(x=x,y=y)) +
  geom_point(aes(color=Countries),size=3)+
  scale_colour_manual(name="Countries",
                      values = c("orangered3","gold2","darkorchid3","navyblue","turquoise2","green3","blue","red","black","gray","orange","darkgreen")) +
  geom_abline(intercept = 0, slope = 1, color="gray60")+
    xlab(xlab)+
    ylab(ylab)+
    ggtitle(title)+ 
    theme_set(theme_bw())
}

#Test of the function for 1 variable
##initial values
Countries <- comp_past_climatic_data_1901_1950$Country #color the points by region or country
x <- data_tot_period$bio1_1901_1950 
y <- data_tot_period$bio1_1951_1990
data <- data_tot_period

##results
s <- scatterplot_funct(data_tot_period,data_tot_period$bio1_1901_1950,data_tot_period$bio1_1951_1990,Countries,"fd","g","t")
```

The goal is to compare the values of each variables between periods. We created a loop to perform the *scatterplot_funct* function that we just created.

```{r loop for the scatterplot_funct to compare the two periods}
bioclim_vars <- c("bio1","bio2","bio4","bio7","bio12","bio15","MeanT_summer","MinT_winter","MeanP_spring_summer","MeanP_summer","AHM")# Add all bioclimatic variable names that we will perform the scatter_plot function


# Loop over bioclimatic variables between periods
for (var in bioclim_vars) {
  # Extract the data for the bioclimatic variable for the two period
  data_subset <- data.frame(data_tot_period[, paste0(var, "_1901_1950")], data_tot_period[, paste0(var, "_1951_1990")]) #for each variable in bioclim_vars, we extract the two periods and created a data.frame with only these two variables. 
  colnames(data_subset) <- c("var1", "var2")
  
  # Create scatterplot to compare the bioclimatic variable between the two periods
  plot <- scatterplot_funct(data_subset, x = data_subset$var1, y = data_subset$var2, Countries,xlab=paste0(var,"_1901_1950"),ylab=paste0(var,"_1951_1990"),title=paste("Comparison of",var,"between the 2 periods"))
  # Print the plot
  print(plot)
  # Add any additional customization or saving the plot if needed
 ggsave(file = paste0(var, "_comparison_plot_raster.png"))
 
}
#Test of the function for 1 variable
##initial values
Countries <- comp_past_climatic_data_1901_1950$Country #color the points by region or country
x <- data_tot_period$bio1_1901_1950 
y <- data_tot_period$bio1_1951_1990
data <- data_tot_period

##results
s <- scatterplot_funct(data_tot_period,data_tot_period$bio1_1901_1950,data_tot_period$bio1_1951_1990,Countries,xlab="Bio1 1901-1950",ylab="Bio1 1951-1990",title="Comparison bio1 between the two periods")
plot(s)
```
Interpretation: **We can see a warming between the period with warmer temperature in the 1951_1990 period. The temperature annual range was more important during the 1901_1950 period -> can be explain by the fact that the winter temperature are more warmer than before in 1950_1990. The annual precipitation vary between the regions as expected by climate change, the precipitation become less common in certain region and more abundant in others. The seasonnality of the precipitation was more important in the first period than in the most recent.Increase in summer precipitation (weird?). Finally, the aridity is increasing over time. 
Answer: **We select the period 1901_1950 because we can see with bio1 or meanT_summer or minT_winter that we have a warming between the period with warmer temperature in the period 1951_1990. We also have an increase in the aridity over time. As the goal is to select the climate of establishment not disturb by human activities, we selected the period 1901_1950 as past period for downstream analysis**

#####################################
Comparison layer of points vs raster#
#####################################
One of the other main issues is that we have 2 types of data with different accuracy: 
    - Layer points data: they have greater resolution but not too convenient for large dataset (mostly for populations, common garden...)
    - Raster data: lower resolution but convenient for large data set (interpolation across the studied area)
  
During the study, we would calculate indexes with values from layer points and raster. We want to test if the values are similar and can be interchangeably applied. 
To do so, we calculated the difference between the values from layer of points and raster for the same populations/pixel.

```{r load raster data}
library(raster)
ras.bio <- stack(list.files("C:/Users/tfrancisco/Documents/Thèse/Data/Climatic/1901-1950/", pattern = ".tif$", full.names = T)) #Permet de combiner l'ensemble des donn?es dans un seul raster
names(ras.bio) <- unlist(strsplit(unlist(lapply(strsplit(list.files("C:/Users/tfrancisco/Documents/Thèse/Data/Climatic/1901-1950/", pattern = ".tif$", full.names = T), split = "C:/Users/tfrancisco/Documents/Thèse/Data/Climatic/1901-1950/"), function(x) x[2])), split = ".tif"))


#we extracted the climatic values for each populations from the raster based on their coordinates
coords<- data.frame(apply(meta_data[,c(5:4)], 2, as.numeric))#we need to have longitude then latitude not the other way around.
past_climatic_data_raster <- data.frame(meta_data$Population,raster::extract(ras.bio, coords))#attention si le package tidyr est actif, cela masque la fonction extract du package raster
colnames(past_climatic_data_raster)=c("Population",names(ras.bio))

```

```{r computation_climatic_variables for raster data }

data_raster <- past_climatic_data_raster #input file 
selected_bio_raster <- c("bio1","bio2","bio4","bio7","bio12","bio15","AHM")#all the variables that are already in the input file (do not need to be calculated) and need to be in the output file
selected_calculated_raster<- c("MeanT_summer","MinT_winter","MeanP_spring_summer","MeanP_summer")#all the variables that are calculated by the function and need to be in the output file

raster_climatic_data <- computation_climatic_variables(data_raster,selected_bio_raster,selected_calculated_raster,mean=T,bio=T)

```


```{r computation_climatic_variables for layer data}
data_layer <- Pre_past_data_1901_1950 #input file
ID_layer <- Pre_past_data_1901_1950$ID#id to calculate the average
from_variable <- 6 #position in the input file of the prec,tmn and tmx variables
to_variable <- 41  #position in the input file of the prec,tmn and tmx variables
selected_bio_layer <- c("bio1","bio2","bio4","bio7","bio12","bio15")#all the variables that have been calculated by the function as bio and need to be in the output file
selected_calculated_layer <- c("MeanT_summer","MinT_winter","MeanP_spring_summer","MeanP_summer","AHM")#all the variables that are calculated by the function and need to be in the output file

past_climatic_data_layer <- computation_climatic_variables(data_layer,selected_bio_layer,selected_calculated_layer,mean=F,bio=F,6,41,ID_layer)
```

Depending on what period we chose, we need to change the name of the variables to fit with the new comparison (we need to change the name of the period for layer of points and raster)
```{r transform the name of the layer of points dataframe and merge dataframes}
#if this is the period 1901_1950 then
past_climatic_data_layer_comp <- merge(meta_data[,c(1:2)],past_climatic_data_layer,"Population")#we only select the retaining variables bio1,2,4,12,15 (and GDD5 but we do not have then now from layer of points)
colnames(past_climatic_data_layer_comp)[3:13] <- paste0(colnames(past_climatic_data_layer_comp)[3:13], "_layer_of_points")

##raster data for period 1901-1950
past_climatic_data_raster_comp <- merge(meta_data[,c(1:2)],raster_climatic_data,"Population")
colnames(past_climatic_data_raster_comp)[3:13] <- paste0(colnames(past_climatic_data_raster_comp)[3:13], "_raster")

#general data frame with the 2 methods
data_tot_period <- data.frame(merge(past_climatic_data_layer_comp,past_climatic_data_raster_comp,"Population")) %>% 
  dplyr::select(-c("Country.y")) 

```


```{r comparision between layer points and raster}
#we will be able to run this when we will have the raster data.
#We will perform the same function as before to compare both of them but the loop will be different
bioclim_vars <- c("bio1","bio2","bio4","bio7","bio12","bio15","MeanT_summer","MinT_winter","MeanP_spring_summer","MeanP_summer","AHM")# Add all bioclimatic variable names that we will perform the scatter_plot functionbioclim_vars <- c("bio1","bio2","bio4","bio7","bio12","bio15","MeanT_summer","MinT_winter","MeanP_spring_summer","MeanP_summer","AHM")# Add all bioclimatic variable names that we will perform the scatter_plot function
# Loop over bioclimatic variables over layer of points vs raster
for (var in bioclim_vars) {
  # Extract the data for the bioclimatic variable for the two periods
  # Extract the data for the bioclimatic variable for the two period
  data_subset <- data.frame(data_tot_period[, paste0(var, "_layer_of_points")], data_tot_period[, paste0(var,"_raster")]) #for each variable in bioclim_vars, we extract the two periods and created a data.frame with only these two variables. 
  colnames(data_subset) <- c("var1", "var2")
  
  # Create scatterplot to compare the bioclimatic variable between the two periods
  plot <- scatterplot_funct(data_subset, x = data_subset$var1, y = data_subset$var2, Countries,xlab=paste0(var,"_layer_of_points"),ylab=paste0(var,"_raster"),title=paste("Comparison of",var,"between the 2 methods"))
  # Print the plot
  print(plot)
  # Add any additional customization or saving the plot if needed
 ggsave(file = paste0(var, "_comparison_plot_methods.png"))
 
}
```

Interpretation: 
Answer: 


    **3. Selection of climatic variables**
    
Selection of climatic variables will be in 2/3 steps: 

      i. Difference between past and future values
      
      
#################################
#           A faire             #
#################################




We look at the difference between past and future values to see if changes will happen in the future for the climatic variables. 
Indeed, if the values won't change, it's not useful to take them into account because it will not affects the populations into the future.
(We could also think that if the climatic values for a specific climatic variables is quit homogeneous along the area of the species, it's not relevant because we might not expect difference in adaptation for this specific variable among populations.)

To investigate whether the values will change, we created a function: *relative_difference_climate* that is creating a new table with the difference between future and present for each specific coordinates populations for each climatic variables(e.g: (tmx08 future - tmx08 present)/tmx08present)

To do so, we first need to process the future climatic data the same way we did for the past data
```{r processing future climatic data}
#load future climatic data
##period 2040_2070
Pre_past_climatic_data_point <- read.csv("C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/Climatic_data/Test_script/pre_future_climatic_data.csv",h=T,sep=";",dec=",")

##Calculate the climatic variables

data_future <- Pre_past_climatic_data_point #input file
ID_future <- Pre_past_climatic_data_point$ID#id to calculate the average
from_variable <- 6 #position in the input file of the prec,tmn and tmx variables
to_variable <- 41  #position in the input file of the prec,tmn and tmx variables
selected_bio_future <- c("bio1","bio2","bio4","bio7","bio12","bio15")#all the variables that have been calculated by the function as bio and need to be in the output file
selected_calculated_future <- c("MeanT_summer","MinT_winter","MeanP_spring_summer","MeanP_summer","AHM")#all the variables that are calculated by the function and need to be in the output file

future_climatic_data <- computation_climatic_variables(data_future,selected_bio_future,selected_calculated_future,mean=F,bio=F,6,41,ID_future)


#add the country information
comp_Future_climatic_data_2040_2070 <- merge(meta_data[,c(1:2)],future_climatic_data,"Population")
colnames(comp_Future_climatic_data_2040_2070)[3:13] <- paste0(colnames(comp_Future_climatic_data_2040_2070)[3:13], "_2041_2070")

##Past period 1901_1950
# Past period = data.frame: comp_past_climatic_data_1901_1950 

```


```{r change in values of variable between present and future, eval=FALSE, include=FALSE}

# Calculate the difference for each climatic variable
relative_difference_climate <- function(data_present, data_future, from,to,Population,number_retained_factor) {
  climatic_columns <- names(data_present)[from:to]#select the climatic variables
  
  relative_data_difference <- (data_future[, climatic_columns] - data_present[, climatic_columns])/data_present[,climatic_columns] #calculate the difference between future values and current one for each climatic variables
  colnames(relative_data_difference) <- paste0(colnames(relative_data_difference),"_relative_difference") #add difference in the name of the columns
  
  ID_long_lat <- data_present %>% 
    group_by(Population) %>% #groups the row by ID
    slice_head(n = 1) %>% #keep only the first row of each groups
    dplyr::select(number_retained_factor)#select only the 3 first columns
  
  #add the ID, longitude and latitude variables
   data_difference_final <- data.frame(ID_long_lat[,number_retained_factor],relative_data_difference) #-1 because ID is already in
   
  return(data_difference_final)
} 
 
data_present <- comp_past_climatic_data_1901_1950#data with the current/past climatic conditions
data_future <- comp_Future_climatic_data_2040_2070 #data with the future climatic conditions
from <- 3 #first columns of climatic data
to <- 14#last columns of climatic data 
Population <- comp_past_climatic_data_1901_1950$Country #pop groups 
number_retained_factor <- c(1,2) 
relative_data_difference<- relative_difference_climate(data_present,data_future,from,to,Population,number_retained_factor)

```

We compare the difference using Violin graphs.
To obtain the violin graph I used a manual code (not run nor show) and then created a function name *violin_plot*.
This graph enables to visualize for each climatic variable the range of difference for each regions.
If a climatic variable has low change (the distance is near 0), the variables is not changing a lot in the future so we could discard it.

```{r graphic representation climatic difference between future and past, eval=FALSE, include=FALSE}

violin_plot <- function(revelant_variables_graph,ID,regions){#may be interesting to do it by regions and not populations to better vizualise
data_violin <- gather(revelant_variables_graph,key="revelant_variables",value="Value",-Population,-Country)#create a dataframe for the violin graph with all the values of climatic variables in 1 columns: Value and all the different coordinates for populations for the different variables in row.
data_violin$Country=as.factor(data_violin$Country)

p <- ggplot(data_violin, aes(x = revelant_variables, y = Value)) +
  geom_violin(trim = FALSE) +
  geom_point(aes(color = Country), shape = 16, position = position_jitter(seed = 1, width = 0.2)) +
  labs(colour="Regions",x="climatic variables",y="Difference between future and past climate",title="Violin Plot of Climatic Factors by Regions")+
  theme(plot.title = element_text(hjust = 0.5))
return(p)
}


revelant_variables_graph <- relative_data_difference#selection of ID, populations and the climatic variables
Groups <- past_climatic_data_1901_1950$Population
regions <-past_climatic_data_1901_1950$Country
  
s=violin_plot(revelant_variables_graph,ID,regions)

#finaly we select only certain data by doing: 
#present
data_present_2step <- subset(data_present_1step,select= -c(...))
  
#future
data_future_2step <- subset(data_future_1step,select= -c(...))
```
Here we can see that the one population has a important difference between the current and the future values so the plot is kinda messy. 
We can decide to do the graph separately for each climatic variable. This will also enable to compare the different future prediction from different climatic models (I do not have yet this kind of information)

```{r violin plot in a loop}
#do the violin plot in a loop for each climatic variables

#input data for the loop
data <- relative_data_difference
regions <-relative_data_difference$Country
colnames(relative_data_difference)=c("Population","Country","bio1","bio2","bio3","bio4","bio5","bio10","bio11","bio12","bio13","bio14","bio15","AHM")
bioclim_vars <- c("bio1", "bio2", "bio3", "bio4","bio5","bio10","bio11","bio12","bio13","bio14","bio15","AHM")

#results: the loop
for (var in bioclim_vars) {
  data_subset <- data.frame(revelant_variables_graph[, paste0(var)]) #for each variable in bioclim_vars, we extract the two periods and created a data.frame with only these two variables. 
  colnames(data_subset) <- c("var1")
  # Create scatterplot to compare the bioclimatic variable between the two periods
  plot <- ggplot(data_subset, aes(x =var,y= var1)) +#var is the values and var1 is just the name of the variable 
  geom_violin(trim = FALSE) +
  geom_point(aes(color = regions), shape = 16, position = position_jitter(seed = 1, width = 0.2)) +
  labs(colour="Regions",x="climatic variables",y=paste0(var,"difference between future and past climate"),title="Violin Plot of Climatic Factors by Regions")+
  theme(plot.title = element_text(hjust = 0.5))
  # Print the plot
  print(plot)
  # Add any additional customization or saving the plot if needed
 
  ######"
    #If we want to add values from different climatic models, we need to do a thing like thats: data_violin <- gather(revelant_variables_graph,key="revelant_variables",value="Value",-Population,-Country)#create a dataframe for the violin graph with all the values of climatic variables in 1 columns: Value and all the different coordinates for populations for the different variables in row.
  #the different values of bio1 for example from different models should be in the same dataframe -> 1 data frame per bioclimatic variables -> maybe we can do a loop for that too. 
} 

```
If, we want to add the values for bio1 and other bio from different climatic models, we need to do like in the function violin plot by creating a columns with the nameof the different models and the values in the same columns. 

Interpretation:...
Conclusion: ...


#################################
#           A faire             #
#################################

      ii. Imprecision along studied area
Accuracy of the climatic variables is not homogeneous along the studied area and also along the climatic variables.
If layers of uncertanty are available, we could check if some variables have a high inaccuracy along the studied area and then discard them. 





```{r Imprecision along the studied area}

```
Interpretation:
Conclusion:

      iii. Multicollinerity between variables
The last step to select the variables is to avoid over collinearity between variables.
To do so, we decided to keep only the variables not too correlated (<absolute corr of 0.7?) from the remaining variables after step 1.

```{r multicollinearity between variables, eval=FALSE, include=FALSE}
#function to do it
#matrix of correlation
correlation_function <-function(data,threshold){
    data_correlation <- subset(data,select= -c(Population,Country)) 
      rownames(data_correlation) <- data$Populations
      correlation <- cor(data_correlation)
  correlation[abs(correlation) <= threshold] <- 0
corr_plot <- corrplot(correlation, method = "number", addrect = 2, col = c("red", "white", "red"), type = "lower", tl.col = "black", tl.cex = 0.6, number.cex = 0.6)
}



#correlation past/present
data_present <- comp_past_climatic_data_1901_1950 #we use the dataset without the variables we discarded previously
threshold <- 0.75
correlation_past <-correlation_function(data_present,threshold)

#correlation future
data_future <- data_future_2step
threshold <- 0.7
correlation_fut <-correlation_function(data_future,threshold)

```
Interpretation:
Conclusion:

```{r retaining variables based on corrplot}

data_past_climatic_data <- comp_past_climatic_data_1901_1950[,c("Population","Country","bio1_1901_1950","bio2_1901_1950","bio4_1901_1950","bio12_1901_1950","bio15_1901_1950")] #add the GDD5
data_future_climatic_data_2040_2070 <- comp_Future_climatic_data_2040_2070[,c()]
```

      iiii. VIF
      
   
   
#################################
#           A faire             #
################################# 




Finally, we tested for the variant inflation factor (VIF) using **...** function to confirm that there is no over collinearity between the remaining variables. To do so, we need the genotyic data

```{r VIF}
#to perfom a vif we need the genotypic data and to perfom a linear regression between the genetic data and all the remaining climatic variables
model_data <- cbind(genotypic_data, comp_past_climatic_data_1901_1950[,c(3:14)])

# Fit the linear regression model
model <- lm(genotypic_data ~ ., data = model_data)

#we can then perfom the VIF
vif.val <- vif(model) # create vector of VIF values
pdf(paste(path1,codesp[sp],"_SamplingLocationsEnvDatExtracted_VIF_Plot.pdf",sep=""), width=8.27*0.6, height=8.27*0.6)
par(oma=c(2,3,0.5,0.5))
barplot(vif.val,  xlab="VIF values", horiz=T, col="gold", las=1, xlim=c(0,10)) # plot horizontal bar charts to display each VIF values
abline(v=5, lwd=3, lty=2)

```
Interpretation: ...
Conclusion: ...


    **4. Comparison of the retained climatic values between populations**

Knowing the climatic range of the sampled populations is important to **have an idea about the selective pressure** that could arise between populations. Also, knowing the repartition of the different regions along the climatic niche is informative regarding the pattern of GEA identified. 
We performed analysis to visualize the climatic niche (distribution of values (not taking into account the populations level but species level) for each climatic variables) and also the violin plot with point represent populations, color represent region to visualize the repartition of populations across range of climatic values for each variable.
We already performed this analysis but now we perfomed it only on the retaining climatic variables
```{r comparison of past climatic values between populations}


#################################################################################################################################
# the real dataframe to use if data_past_climatic_data but for now i do have it so i use the one without removing the variables
#################################################################################################################################


##to compare all the variables in the same graph, we need to scale them or do one graph per values, we will do both:
##1 variable per graph: 
#we already did that in the section: violin plot in a loop
data <- data_past_climatic_data
revelant_variables_graph <- data_past_climatic_data[,-c(1,2)]
regions <-data_past_climatic_data$Country
colnames(data_past_climatic_data)=c("Population","Country","bio1","bio2","bio3","bio4","bio5","bio10","bio11","bio12","bio13","bio14","bio15","AHM")
bioclim_vars <- c("bio1_1901_1950", "bio2_1901_1950", "bio4_1901_1950","bio12_1901_1950","bio15_1901_1950")

#results: the loop
for (var in bioclim_vars) {
  data_subset <- data.frame(revelant_variables_graph[, paste0(var)]) #for each variable in bioclim_vars, we extract the two periods and created a data.frame with only these two variables. 
  colnames(data_subset) <- c("var1")
  # Create scatterplot to compare the bioclimatic variable between the two periods
  plot <- ggplot(data_subset, aes(x =var,y= var1)) +#var is the values and var1 is just the name of the variable 
  geom_violin(trim = FALSE) +
  geom_point(aes(color = regions), shape = 16, position = position_jitter(seed = 1, width = 0.2)) +
  labs(colour="Regions",x="climatic variables",y=paste0(var,"difference between future and past climate"),title="Violin Plot of Climatic Factors by Regions")+
  theme(plot.title = element_text(hjust = 0.5))
  # Print the plot
  print(plot)
  # Add any additional customization or saving the plot if needed
}

##scale 
revelant_variables_scale <- data.frame(scale(comp_past_climatic_data_1901_1950[,-c(1,2)]))# scale the climatic variables of interest

#we using the violin_plot function 
revelant_variables_graph <- data.frame(comp_past_climatic_data_1901_1950[,c(1,2)],revelant_variables_scale)#selection of ID, populations and the climatic variables
Populations <- revelant_variables_graph$Population
regions <-revelant_variables_graph$Country


violin_plot(revelant_variables_graph,ID,regions)


```
Interpretations: 


#################################
#           A faire             #
#################################
We also did it for the future:
```{r comparison of future climatic values between populations}

#################################################################################################################################
# the real dataframe to use if data_past_climatic_data but for now i do have it so i use the one without removing the variables
#################################################################################################################################

#it's the same thing that above but for future data

##1 variable per graph: 

data <- comp_future_climatic_data_2040_2070
revelant_variables_graph <- comp_future_climatic_data_2040_2070[,-c(1,2)]
regions <-comp_future_climatic_data_2040_2070$Country
colnames(comp_future_climatic_data_2040_2070)=c("Population","Country","bio1","bio2","bio3","bio4","bio5","bio10","bio11","bio12","bio13","bio14","bio15","AHM")
bioclim_vars <- c("bio1", "bio2", "bio3", "bio4","bio5","bio10","bio11","bio12","bio13","bio14","bio15","AHM")

#results: the loop
for (var in bioclim_vars) {
  data_subset <- data.frame(revelant_variables_graph[, paste0(var)]) #for each variable in bioclim_vars, we extract the two periods and created a data.frame with only these two variables. 
  colnames(data_subset) <- c("var1")
  # Create scatterplot to compare the bioclimatic variable between the two periods
  plot <- ggplot(data_subset, aes(x =var,y= var1)) +#var is the values and var1 is just the name of the variable 
  geom_violin(trim = FALSE) +
  geom_point(aes(color = regions), shape = 16, position = position_jitter(seed = 1, width = 0.2)) +
  labs(colour="Regions",x="climatic variables",y=paste0(var,"difference between future and past climate"),title="Violin Plot of Climatic Factors by Regions")+
  theme(plot.title = element_text(hjust = 0.5))
  # Print the plot
  print(plot)
  # Add any additional customization or saving the plot if needed
}



##scale 
revelant_variables_scale <- data.frame(scale(comp_future_climatic_data_2040_2070[,-c(1,2)]))# scale the climatic variables of interest

#we using the violin_plot function 
revelant_variables_graph <- data.frame(comp_future_climatic_data_2040_2070[,c(1,2)],revelant_variables_scale)#selection of ID, populations and the climatic variables
Populations <- revelant_variables_graph$Population
regions <-revelant_variables_graph$Country 


violin_plot(revelant_variables_graph,ID,regions)

```
Interpretations: 

    **5. Comparison of the future retained climatic values between populations**
    
Multiple climatic models that predict the future climate are available. Choosing one of them could be a challenge because sometimes the difference in prediction are important. To avoid bias by selecting only one climatic model that could be an extreme, we selected 3 or 5 of them for the future data. 
We want to compare them to see if there is a lot a differences in predictions between models. 
To do so, we performed a ... see paper of benjamin or felix showing that
```{r Comparison of the climatic models prediction}
#look above,I explained how we can modify the loop of violin plot to compare the values of predictions for the different climatic models into the future 
```

    **6. Standardization of the retained climatic values**

The final step is to create a data frame with the retained climatic variables standardized for further analysis. 
We standardized the past and future data with the same values to enable comparison of index between past and future based on the climatic data (e.g Adaptive index, genomic offset...)

```{r normalize the data, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}
#creation of the scaled matrix


#initial parameters
data <- data_past_climatic_data[,-c(1,2)]
ID <- data_past_climatic_data[,c(1,2)]
scale_f(data,ID)

##scale the selected variables then add ID and reorganize the columns order
  data_scale <- scale(data) %>% #scale
    data.frame(ID) %>% #add ID
    dplyr::select(7,8,1:6)#reorganize the column order
  
  #we kept the values of scaling and centering for each variables to normalize the future values by the same values
  scale_env_value <- attr(data_scale, 'scaled:scale')
center_env_value <- attr(data_scale, 'scaled:center')

#Save the matrix and the normalized values

write_xlsx(data_scale,"C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Pinus_pinea/normalized_past_climatic_data.xlsx")
write_xlsx(scale_env_value,"C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Pinus_pinea/scale_env_value.xlsx")
write_xlsx(center_env_value,"C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Pinus_pinea/center_env_value.xlsx")

```
