---
title: "Variance_Partitionning_ind_level"
author: "Thomas Francisco"
date: "2024-04-01"
output: html_document
---


```{r setup, include=FALSE}
rm(list = ls())
knitr::opts_chunk$set(cache = FALSE)

library(vegan)
library(dplyr)
library(robust) #for rdadapt
library(qvalue) #for rdadapt


```




```{r intial data}

#climatic data
##0.6 corr
Past_climatic_0.6corr <- read.csv("C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/Climatic_data/corr0.6/Past_Climatic_data_scale_0.6cor.csv",sep=";",dec=",")

#0.75corr
Past_climatic_0.75corr <- read.csv("C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/Climatic_data/corr0.75/Past_Climatic_data_scale_0.75cor.csv",sep=";",dec=",")


#Population Structure
#PCs axis
structure_PCs <- read.csv("C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/GEA/Structure_proxy/PCa_score_T_adapcon_gentree_scale.csv",sep=";",dec=",")

#STRUCTURE Groups
load("C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/GEA/Structure_proxy/Cluster_score_STRUCTURE_T_Adapcon_gentree.Rdata")
Structure_cluster <- Cluster_score_STRUCTURE_T_Adapcon_gentree

#IBD proxy
#dbMEMs
IBD_dbMEMs <- read.csv("C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/GEA/dbMEMs/meta_data_dbmem_29pop_adapcon_gentree_scale.csv",sep=";",dec=",")


#genomic data
load("C:/Users/tfrancisco/Documents/Thèse/Data/Espèces/Taxus_baccata/genetic_new/data_allelic_frequencies_29pop_adapcon_gentree_475_8616.Rdata")

genomic_matrix <- data_allelic_frequencies_29pop_adapcon_gentree_475_8616

```


```{r merge all informations into 1 table}


#0.75 corr
Data_RDA <- merge(Past_climatic_0.75corr,structure_PCs,"Population") %>% merge(Structure_cluster,"Population") %>% merge(IBD_dbMEMs,"Population") 
Data_tot_RDA <- Data_RDA[,-c(2,10,11,13,16)] %>% dplyr::select(c(1,9,2:8,10:15))

Data_tot_RDA <- Data_RDA %>% dplyr::select("Population","Country","Annual_Tc","Tc_diur_range","Tc_Seasonality","Annual_P","P_Seasonality","PC1","PC2","Group1","Group2","Group3","Longitude","Latitude","MEM1","MEM2","MEM3","MEM4")


scale_coord <- Data_tot_RDA[,c(13,14)] %>%
  apply(2,as.numeric) %>% as.data.frame %>% 
mutate(across(where(is.numeric), scale))

data_tot_RDA_final <- data.frame(Data_tot_RDA[,-c(13,14)],scale_coord)
#0.6 corr
#Data_RDA <- merge(Past_climatic_0.6corr,structure_PCs,"Population") %>% merge(Structure_cluster,"Population") %>% merge(IBD_dbMEMs,"Population") 
#Data_tot_RDA <- Data_RDA %>% dplyr::select("Population","Country","Annual_Tc","Tc_diur_range","Tc_Seasonality","Annual_P","P_Seasonality","PC1","PC2","Group1","Group2","Group3","Longitude","Latitude","MEM1","MEM2","MEM3","MEM4")

test_list <- list(genomic_data=genomic_matrix,
  climatic_variables= data_tot_RDA_final[,c("Annual_Tc","Tc_diur_range","Tc_Seasonality","Annual_P","P_Seasonality")],
                  Group_structure=data_tot_RDA_final[,c("Group1","Group2","Group3")],
                  PCs_Structure=data_tot_RDA_final[,c("PC1","PC2")],
                  IBD_dbMEms=data_tot_RDA_final[,c("MEM1","MEM2","MEM3","MEM4")],
                  IBD_coordinates=data_tot_RDA_final[,c("Longitude","Latitude")])


```

One assumption of the GEA models is that the environmental variables explains some of the genetic variation. We want to test that, we realize a variance partitioning. 
Also, the variance partitioning enables to visualize if there is a lot a covariation between the structure, IBD and environment 
```{r variance partitionning with dbMEMs as IBD proxy}


#total model
M_tot_db <- rda(formula=genomic_matrix~Annual_Tc+Tc_diur_range+Tc_Seasonality+Annual_P+P_Seasonality+PC1+PC2+MEM1+MEM2+MEM3+MEM4,data = Data_tot_RDA,scale=T)
RsquareAdj(M_tot_db)
anova.cca(M_tot_db)


##partial_ancestry_model
M_ancestry_db <- rda(formula = genomic_matrix ~ PC1+PC2 +Condition(Annual_Tc+Tc_diur_range+Tc_Seasonality+Annual_P+P_Seasonality +MEM1+MEM2+MEM3+MEM4), data = Data_tot_RDA,scale=T)
RsquareAdj(M_ancestry_db)
anova.cca(M_ancestry_db)


#partial_geo_model
M_geo_db <- rda(formula = genomic_matrix ~ MEM1+MEM2+MEM3+MEM4+ Condition(Annual_Tc+Tc_diur_range+Tc_Seasonality+Annual_P+P_Seasonality+PC1+PC2), data = Data_tot_RDA,scale=T)
RsquareAdj(M_geo_db)
anova.cca(M_geo_db)



#partial_env_RDA
M_IBE_db <- rda(formula = genomic_matrix ~  Annual_Tc+Tc_diur_range+Tc_Seasonality+Annual_P+P_Seasonality +Condition(PC1+PC2+ MEM1+MEM2+MEM3+MEM4) , data = Data_tot_RDA,scale=T)
RsquareAdj(M_IBE_db)
summary(M_IBE_db)
anova.cca(M_IBE_db)

```
Interpretation: 

```{r variance partitionning with coord as IBD proxy}

#to perfom the variance partioning with the coord as IBD proxy, we need to scale the coords


#total model
M_tot_coord <- rda(formula=genomic_matrix~Annual_Tc+Tc_diur_range+Tc_Seasonality+Annual_P+P_Seasonality+PC1+PC2+Longitude+Latitude,data = data_tot_RDA_final,scale=T)
RsquareAdj(M_tot_coord)
anova.cca(M_tot_coord)


##partial_ancestry_model
M_ancestry_coord <- rda(formula = genomic_matrix ~ PC1+PC2 +Condition(Annual_Tc+Tc_diur_range+Tc_Seasonality+Annual_P+P_Seasonality +Longitude+Latitude), data = data_tot_RDA_final,scale=T)
RsquareAdj(M_ancestry_coord)
anova.cca(M_ancestry_coord)


#partial_geo_model
M_geo_coord <- rda(formula = genomic_matrix ~ Longitude+Latitude+ Condition(Annual_Tc+Tc_diur_range+Tc_Seasonality+Annual_P+P_Seasonality+PC1+PC2), data = data_tot_RDA_final,scale=T)
RsquareAdj(M_geo_coord)
anova.cca(M_geo_coord)



#partial_env_RDA
M_IBE_coord <- rda(formula = genomic_matrix ~  Annual_Tc+Tc_diur_range+Tc_Seasonality+Annual_P+P_Seasonality +Condition(PC1+PC2+Longitude+Latitude) , data = data_tot_RDA_final,scale=T)
RsquareAdj(M_IBE_coord)
summary(M_IBE_coord)
anova.cca(M_IBE_coord)

```
Interpretation:


Comparison between dbMEM and coord as IBD proxies: 


Comparison between PCs and Score Structure as Structure proxies:

```{r variance partitionning with mscore_structure as Structure proxy and dbMEMs as IBD proxy}


#total model
M_tot_db <- rda(formula=genomic_matrix~Annual_Tc+Tc_diur_range+Tc_Seasonality+Annual_P+P_Seasonality+Group1+Group2+Group3+MEM1+MEM2+MEM3+MEM4,data = data_tot_RDA_final,scale=T)
RsquareAdj(M_tot_db)
anova.cca(M_tot_db)


##partial_ancestry_model
M_ancestry_db <- rda(formula = genomic_matrix ~ Group1+Group2+Group3 +Condition(Annual_Tc+Tc_diur_range+Tc_Seasonality+Annual_P+P_Seasonality +MEM1+MEM2+MEM3+MEM4), data = data_tot_RDA_final,scale=T)
RsquareAdj(M_ancestry_db)
anova.cca(M_ancestry_db)


#partial_geo_model
M_geo_db <- rda(formula = genomic_matrix ~ MEM1+MEM2+MEM3+MEM4+ Condition(Annual_Tc+Tc_diur_range+Tc_Seasonality+Annual_P+P_Seasonality+Group1+Group2+Group3), data = data_tot_RDA_final,scale=T)
RsquareAdj(M_geo_db)
anova.cca(M_geo_db)



#partial_env_RDA
M_IBE_db <- rda(formula = genomic_matrix ~  Annual_Tc+Tc_diur_range+Tc_Seasonality+Annual_P+P_Seasonality +Condition(Group1+Group2+Group3+ MEM1+MEM2+MEM3+MEM4) , data = data_tot_RDA_final,scale=T)
RsquareAdj(M_IBE_db)
summary(M_IBE_db)
anova.cca(M_IBE_db)

```


```{r variance partitionning with coord as IBD proxy}

#to perfom the variance partioning with the coord as IBD proxy, we need to scale the coords


#total model
M_tot_coord <- rda(formula=genomic_matrix~Annual_Tc+Tc_diur_range+Tc_Seasonality+Annual_P+P_Seasonality+Group1+Group2+Group3+Longitude+Latitude,data = data_tot_RDA_final,scale=T)
RsquareAdj(M_tot_coord)
anova.cca(M_tot_coord)


##partial_ancestry_model
M_ancestry_coord <- rda(formula = genomic_matrix ~ Group1+Group2+Group3 +Condition(Annual_Tc+Tc_diur_range+Tc_Seasonality+Annual_P+P_Seasonality +Longitude+Latitude), data = data_tot_RDA_final,scale=T)
RsquareAdj(M_ancestry_coord)
anova.cca(M_ancestry_coord)


#partial_geo_model
M_geo_coord <- rda(formula = genomic_matrix ~ Longitude+Latitude+ Condition(Annual_Tc+Tc_diur_range+Tc_Seasonality+Annual_P+P_Seasonality+Group1+Group2+Group3), data = data_tot_RDA_final,scale=T)
RsquareAdj(M_geo_coord)
anova.cca(M_geo_coord)



#partial_env_RDA
M_IBE_coord <- rda(formula = genomic_matrix ~  Annual_Tc+Tc_diur_range+Tc_Seasonality+Annual_P+P_Seasonality +Condition(Group1+Group2+Group3+Longitude+Latitude) , data = data_tot_RDA_final,scale=T)
RsquareAdj(M_IBE_coord)
summary(M_IBE_coord)
anova.cca(M_IBE_coord)

```

Interpretation: 
Conclusion: we will retain the model with dbMEMs as IBD and cluster score as population structure proxy


The next step of the analysis is to perform GEA outliers detection to see if we can detect loci link to climatic conditions. 
We will perform this analysis correction or not by population structure.

In this script, we will perform RDA(not accounting for population structure) and pRDA (accounting for population structure):
**explain principle of RDA**

        -RDA without correcting

We perform the RDA without correcting
```{r RDA without correcting, message=FALSE, warning=FALSE}

RDA_env <- rda(formula = genomic_matrix ~  Annual_Tc+Tc_diur_range+Tc_Seasonality+Annual_P+P_Seasonality, data = data_tot_RDA_final,scale=T)
RsquareAdj(RDA_env)
summary(RDA_env)
anova.cca(RDA_env)
```
Then, we need to perform the outliers selection on this models. To do this, we will follow procedure in Capblancq et al. 2021 and we will identified outliers based on their extremeness along a distribution of Mahalanobis distances estimated between each locus and the center of the RDA space using a certain number of axes (K). 

First, we need to select the number of ordination axis (RDA axis) that we will retains
```{r screeplot}
#screeplot
plot(RDA_env$CCA$eig, option="screeplot")
#or like that
screeplot(RDA_env)

#explained variance along each RDA axis
RDA_env$CCA$eig

vif.cca(RDA_env)

```
Interpretation: based on this, we will keep the 2 first RDA axis to perform the outliers detection


We can perform the genome scan procedure to calculate the Mahalanobis distance and then qvalues and pvalues associate to this distance to do a selection of outliers 
```{r genome scan function}
rdadapt<-function(rda,K)
{
  zscores<-rda$CCA$v[,1:as.numeric(K)]
  resscale <- apply(zscores, 2, scale)
  resmaha <- covRob(resscale, distance = TRUE, na.action= na.omit, estim="pairwiseGK")$dist
  lambda <- median(resmaha)/qchisq(0.5,df=K)
  reschi2test <- pchisq(resmaha/lambda,K,lower.tail=FALSE)
  qval <- qvalue(reschi2test)
  q.values_rdadapt<-qval$qvalues
  return(data.frame(p.values=reschi2test, q.values=q.values_rdadapt))
}

#Perform the fucntion to calculate the mahalanobis distance and then pvalues/ qvalues

genome_scan_RDA <- rdadapt(RDA_env,2)

plot(RDA_env, scaling=3) 
```

Now, we need to select a threshold and then save the outlier
```{r selection of outliers}
#threshold qvalues < 0.05

thres_env <- 0.1/length(genome_scan_RDA$p.values)

outliers <- data.frame(Loci = colnames(genomic_matrix)[which(genome_scan_RDA$p.values<thres_env)], p.value = genome_scan_RDA$p.values[which(genome_scan_RDA$p.values<thres_env)], contig = unlist(lapply(strsplit(colnames(genomic_matrix)[which(genome_scan_RDA$p.values<thres_env)], split = "_"), function(x) x[1])))

outliers
```


We also perform the outlier detection based on the loadings of loci along the RDa axis to ssearch for extreme loadings on each retained axis
```{r extremness along RDA axis}
load.rda <- scores(RDA_env, choices=c(1:3), display="species")

hist(load.rda[,1], main="Loadings on RDA1")
hist(load.rda[,2], main="Loadings on RDA2")

```
```{r perform the detection }
#function to do it
outliers_function <- function(x,z){
  lims <- mean(x) + c(-1, 1) * z * sd(x)     # find loadings +/-z sd from mean loading     
  x[x < lims[1] | x > lims[2]]               # locus names in these tails
}

cand1 <- outliers_function(load.rda[,1],2.5) # 38
cand2 <- outliers_function(load.rda[,2],2.5) # 69


ncand <- length(cand1) + length(cand2)
ncand

```



```{r pRDA correcting for pop structure}

#partial_env_RDA
pRDA_env <- rda(formula = genomic_matrix ~  Annual_Tc+Tc_diur_range+Tc_Seasonality+Annual_P+P_Seasonality +Condition(Group1+Group2+Group3+ MEM1+MEM2+MEM3+MEM4) , data = data_tot_RDA_final,scale=T)
RsquareAdj(pRDA_env)
summary(pRDA_env)
anova.cca(pRDA_env)

```

```{r}
cand1 <- cbind.data.frame(rep(1,times=length(cand1)), names(cand1), unname(cand1))
cand2 <- cbind.data.frame(rep(2,times=length(cand2)), names(cand2), unname(cand2))


colnames(cand1) <- colnames(cand2) <- colnames(cand3) <- c("axis","snp","loading")

cand <- rbind(cand1, cand2, cand3)
cand$snp <- as.character(cand$snp)
```

